# 🔎 Análise Amostra

## Origem dos Dados

Na fase da coleta de dados foi gerado o arquivo artigos_en.csv com 2.010 artigos em inglês. A partir desses dados foi gerado o arquivo artigos_en_amostra.csv com 201 artigos sorteados aleatoriamente (cerca de 10% por subcategoria).

A amostra gerada foi validada manualmente pelo time. Cada artigo foi lido e avaliado considerando as seguintes notas: 

0 = categoria incorreta

1 = categoria parcialmente adequada

2 = categoria correta

As notas foram registradas em planilha do Google Sheets e fornecidas para categoria, subcategoria e tags do texto. Esses dados foram utilizados para definir a qualidade dos dados para o projeto.

## Avaliação da Amostra

Cada registro da amostra foi avaliado individualmente pelos membros da equipe, para facilitar a validação manual, a amostra foi organizada em uma planilha compartilhada, contendo as principais informações de cada artigo, como id, categoria, subcategoria, título, texto, tags, url e Obs quando necessário. Cada integrante da equipe identificava seu nome na linha do registro que escolheu validar e a partir da URL realizava a leitura do artigo. Após essa análise, cada integrante registrava as notas de Texto x Categoria.
Avaliação das tags e subcategorias 

As subcategorias e as tags também foram avaliadas individualmente em relação à categoria atribuída. As subcategorias são utilizadas como parâmetros de busca na API do DEV.to, que utiliza as tags atribuídas pelos autores para localizar e retornar os artigos. Dessa forma, a avaliação humana permitiu verificar a qualidade dos artigos recuperados por cada busca: caso uma tag apresentasse baixo desempenho ou retornasse conteúdos pouco relacionados à categoria esperada, sua busca poderia ser descartada e substituída por outra subcategoria/tag, buscando recuperar textos mais adequados à categoria utilizada no treinamento.
Métricas de avaliação 

Para medir a qualidade das categorias, foram calculados os percentuais de classificações corretas e de classificações parcialmente adequadas:

Nota 2: quantidade de registros com nota 2 / quantidade total de registros da categoria × 100
Notas 1 + 2: quantidade de registros com nota diferente de 0 / quantidade total de registros da categoria × 100

**Resultados** 

Os resultados de Texto x Categoria foram comparados entre as seis categorias:

< categoria> (Nota 2 / Nota 1+2)

Backend (78,79% / 93,94%)
Cloud (48,57% / 54,29%)
DataScience (54,84% / 70,97%)
Database (68,75% / 75,00%)
Frontend (68,57% / 77,14%)
Security (85,71% / 97,14%)

No conjunto da amostra, 67,66% dos registros receberam nota 2 e 78,11% receberam nota 1 ou 2.

**Análise das subcategorias** 

A decisão de manter ou eliminar subcategorias considerou tanto os indicadores quantitativos quanto a análise qualitativa das observações registradas durante a validação. Foram identificados, por exemplo:

Cloud/OCI (33,33% de nota 2)
DataScience/AI (50%)
DataScience/Deep Learning (0%)
Frontend/TypeScript (0%)
Database/MongoDB (60%)

Entretanto, uma baixa porcentagem não foi considerada, isoladamente, motivo para eliminação, pois algumas dessas subcategorias ainda contribuem para trazer textos relevantes e representativos para o treinamento do modelo. No caso de IA, por exemplo, foi observado que muitos dos conteúdos retornados tratavam de Inteligência Artificial de forma mais independente de DataScience, portanto, tanto o desempenho percentual quanto a relevância dos registros para a categoria.

**Tratamento dos conflitos**

Durante a análise para a limpeza dos dados foram identificados 71 casos em que o mesmo ID e texto estavam associados a categorias diferentes, decorrentes principalmente de artigos que possuíam múltiplas tags. Para resolver esses conflitos de forma padronizada, foram estabelecidas regras hierárquicas:

* Maior nota de desempenho da subcategoria
* Em caso de empate, menor quantidade de registros associados à subcategoria
* Persistindo o empate, menor quantidade de registros associados à categoria

Essas regras permitiram selecionar a classificação com maior evidência de coerência na validação, reduzindo duplicidades e evitando decisões subjetivas.

**Resultado final da amostra**

A análise identificou subcategorias com menor aderência entre os artigos recuperados e a categoria atribuída. Como resultado, foram eliminadas as subcategorias OCI (Cloud), AI (DataScience) e DeepLearning (DataScience), considerando seu desempenho na validação e a qualidade dos conteúdos recuperados. 

Após esse filtro, a amostra foi considerada adequada para orientar a seleção das subcategorias utilizadas na coleta, permitindo reduzir buscas com menor aderência e priorizar aquelas com maior capacidade de recuperar conteúdos coerentes com suas respectivas categorias.
Após o descarte das subcategorias com menor desempenho na validação da amostra, restaram 1.851 registros, utilizados como entrada para a etapa de limpeza dos dados.

## Objetivo

Este notebook tem como objetivo:

* Descrever os passos realizados na análise da amostra 
* Descrever os resultados obtidos 
* Realizar um filtro na base gerada desconsiderando os registros das categorias | subcategorias que tiveram avaliação insatisfatória no teste da amostra
* Verificar como ficou distribuição das categorias após remoção 
* Gerar a base que será utilizada para Limpeza de Dados e posteriormente para treinar o modelo. 

## Bibliotecas

In [1]:
# Importando bibliotecas
import pandas as pd

## Obtendo Dados

In [2]:
# Obtendo dados
df = pd.read_csv('../dados/artigos_en.csv')
df.head(2)

,categoria,collection_id,comments_count,created_at,crossposted_at,description,edited_at,flare_tag,id,language,...,published_timestamp,readable_publish_date,reading_time_minutes,subcategoria,subforem_id,tag_list,tags,texto,title,type_of
0,Frontend,NaN,0,2026-07-18T19:29:08Z,NaN,Choosing a state management library feels a bi...,2026-07-18T19:31:43Z,NaN,4175536,en,...,2026-07-18T19:29:08Z,Jul 18,3,frontend,1.0,"['architecture', 'frontend', 'react', 'typescr...","architecture, frontend, react, typescript",Choosing a state management library feels a bi...,Stop Over Engineering Your Frontend State: A D...,article
1,Frontend,NaN,0,2026-07-16T15:15:36Z,NaN,"Learn what a design system is, why modern engi...",NaN,NaN,4159443,en,...,2026-07-18T05:00:00Z,Jul 18,5,frontend,1.0,"['frontend', 'architecture', 'design', 'system...","frontend, architecture, design, systemdesign","\r\n> Learn what a design system is, why moder...",Design Systems Explained,article


## Conferindo Dados

In [3]:
# Confirmando estrutura e quantidade dos dados 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2010 entries, 0 to 2009
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   categoria                 2010 non-null   object 
 1   collection_id             205 non-null    float64
 2   comments_count            2010 non-null   int64  
 3   created_at                2010 non-null   object 
 4   crossposted_at            15 non-null     object 
 5   description               2010 non-null   object 
 6   edited_at                 356 non-null    object 
 7   flare_tag                 88 non-null     object 
 8   id                        2010 non-null   int64  
 9   language                  2010 non-null   object 
 10  last_comment_at           2010 non-null   object 
 11  positive_reactions_count  2010 non-null   int64  
 12  public_reactions_count    2010 non-null   int64  
 13  published_at              2010 non-null   object 
 14  publishe

In [4]:
# Distribuição por categoria
df['categoria'].value_counts()

categoria
Security       346
Cloud          345
Frontend       341
Backend        338
Database       324
DataScience    316
Name: count, dtype: int64

In [5]:
# Distribuição por subcategoria
df['subcategoria'].value_counts()

subcategoria
cryptography       60
infosec            60
security           60
javascript         59
ethicalhacking     59
java               59
azure              59
api                59
database           59
aws                59
nosql              58
cybersecurity      58
python             58
kubernetes         58
datascience        58
ai                 58
oci                58
cloud              58
typescript         58
nextjs             58
csharp             56
postgres           56
react              56
backend            56
css                55
frontend           55
springboot         54
php                54
mongodb            54
docker             53
sql                52
machinelearning    52
mfa                49
pandas             47
crud               45
deeplearning       43
Name: count, dtype: int64

In [6]:
# Selecionando registros da subcategoria deeplearning para análise
df[df['subcategoria'] == 'deeplearning'][['id', 'categoria', 'subcategoria', 'texto', 'tags']].head(5)

,id,categoria,subcategoria,texto,tags
1967,4171806,DataScience,deeplearning,## What Changed\r\nWan-AI has released Wan-Dan...,"aiml, videogeneration, musictodance, deeplearning"
1968,4168478,DataScience,deeplearning,In this new AI world data is one of the most v...,"ai, programming, learning, deeplearning"
1969,4175137,DataScience,deeplearning,{{ $json.postContent }},"ai, deeplearning, computerscience, machinelear..."
1970,4174886,DataScience,deeplearning,{{ $json.postContent }},"ai, deeplearning, computerscience, machinelear..."
1971,4170810,DataScience,deeplearning,# Open Models Now Run 63% of AI's Token Traffi...,"opensource, ai, machinelearning, deeplearning"


In [7]:
# Verificando quantos textos são iguais a {{ $json.postContent }}
(df['texto'] == '{{ $json.postContent }}').sum()

np.int64(42)

## Descarte de subcategorias

In [8]:
# Definindo subcategorias que serão eliminadas 
subcategorias_descartar =  ['oci', 'ai', 'deeplearning']

In [9]:
# Desconsiderando subcategorias
df_filtrado = df[~df['subcategoria'].isin(subcategorias_descartar)]

In [10]:
df_filtrado.shape

(1851, 24)

In [11]:
# Distribuição dos registros por categoria
df_filtrado['categoria'].value_counts()

categoria
Security       346
Frontend       341
Backend        338
Database       324
Cloud          287
DataScience    215
Name: count, dtype: int64

## Gerando Nova Base

In [12]:
# Gerar base para treino e teste
df_filtrado.to_csv("../dados/artigos_en_filtrado.csv", encoding="utf-8-sig", index=False)


Após descarte das subcaregorias com menor desempenho na validação da amostra, restaram 1851 registros. 